lib/fluent/plugin/out_splunk_ingest_api.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+require 'fluent/plugin/out_splunk'
+require 'openid_connect'
+require 'rack/oauth2'
+require 'multi_json'
+
+module Fluent::Plugin
+  class SplunkIngestApiOutput < SplunkOutput
+    Fluent::Plugin.register_output('splunk_ingest_api', self)
+
+    desc 'Service Client Identifier'
+    config_param :service_client_identifier, :string, default: nil
+
+    desc 'Service Client Secret Key'
+    config_param :service_client_secret_key, :string, default: nil
+
+    desc 'Token Endpoint'
+    config_param :token_endpoint, :string, default: '/system/identity/v1/token'
+
+    desc 'Ingest Api Hostname'
+    config_param :ingest_api_host, :string, default: 'api.splunkbeta.com'
+
+    desc 'Ingest API Tenant Name'
+    config_param :ingest_api_tenant, :string
+
+    desc 'Ingest API Events Endpoint'
+    config_param :ingest_api_events_endpoint, :string, default: '/ingest/v1beta2/events'
+
+    desc 'Debug the HTTP transport'
+    config_param :debug_http, :bool, default: false
+
+    def prefer_buffer_processing
+      true
+    end
+
+    def configure(conf)
+      super
+    end
+
+    def construct_api
+      uri = "https://#{@ingest_api_host}/#{@ingest_api_tenant}#{@ingest_api_events_endpoint}"
+      URI(uri)
+    rescue StandardError
+      raise Fluent::ConfigError, "URI #{uri} is invalid"
+    end
+
+    def format(tag, time, record)
+      format_event(tag, time, record)
+    end
+
+    def format_event(tag, time, record)
+      event = prepare_event_payload(tag, time, record)
+      # Unsure how to drop a record. So append the empty string
+      if event[:body].nil? || event[:body].strip.empty?
+        ''
+      else
+        MultiJson.dump(event) + ','
+      end
+    end
+
+    def prepare_event_payload(tag, time, record)
+      payload = super(tag, time, record)
+      payload[:attributes] = payload.delete(:fields) || {}
+      payload[:attributes][:index] = payload.delete(:index) if payload[:index]
+      payload[:body] = payload.delete(:event)
+      payload.delete(:time)
+      payload[:timestamp] = (time.to_f * 1000).to_i
+      payload[:nanos] = time.nsec / 100_000
```

Comment:
can't find that the `id` param is being used. I think this should be used by default to prevent duplicated events in splunk when something fails halfway and fluentd resends the buffer again.

Either a user should be allowed to set a field to be used as unique id or it should be generated automatically and safed into the buffer (as far as i understand `#format`  will be called before saving to the buffer).